{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "638f8f95-2782-45e7-9161-4d8416059ac4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import json\n",
    "from io import StringIO\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from dotenv import dotenv_values\n",
    "from pyexeggutor import (\n",
    "    format_header,\n",
    "    read_pickle,\n",
    ")\n",
    "import matplotlib.colors as mcolors\n",
    "import seaborn as sns\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "\n",
    "\n",
    "from kegg_pathway_profiler.pathways import (\n",
    "    pathway_coverage_wrapper,\n",
    ")\n",
    "from kegg_pathway_profiler.enrichment import (\n",
    "    unweighted_pathway_enrichment_wrapper,\n",
    ")\n",
    "\n",
    "from nichespace.neighbors import (\n",
    "    pairwise_distances_kneighbors,\n",
    "    KNeighborsLeidenClustering,\n",
    ")\n",
    "from nichespace.manifold import (\n",
    "    HierarchicalNicheSpace,\n",
    "    QualitativeSpace,\n",
    "    EmbeddingAnnotator,\n",
    "    DEFAULT_REGRESSOR,\n",
    "    DEFAULT_REGRESSOR_PARAM_SPACE,\n",
    ")\n",
    "\n",
    "\n",
    "from nichespace.llm import (\n",
    "    LLMAnnotator,\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62dabb32-6643-46a5-8d66-a285a408e862",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%time\n",
    "# Data\n",
    "df_quality = pd.read_csv(\"../data/quality.tsv.gz\",sep=\"\\t\", index_col=0)\n",
    "\n",
    "quality_label=\"completeness_gte90.contamination_lt5\"\n",
    "# quality_label=\"completeness_gte50.contamination_lt10\"\n",
    "\n",
    "genome_to_clusterani = pd.read_csv(f\"../data/training/{quality_label}/y.tsv.gz\", sep=\"\\t\", index_col=0, header=None).iloc[:,0].astype(\"category\")\n",
    "X_genomic_traits = pd.read_csv(f\"../data/training/{quality_label}/X.tsv.gz\", sep=\"\\t\", index_col=0).astype(bool)\n",
    "X_genomic_traits_clusterani = pd.read_csv(f\"../data/training/{quality_label}/X_grouped.tsv.gz\", sep=\"\\t\", index_col=0).astype(bool)\n",
    "\n",
    "genome_to_taxonomy = pd.read_csv(\"../data/taxonomy.tsv.gz\", sep=\"\\t\", index_col=0).iloc[:,0]\n",
    "clusterani_to_taxonomy = pd.read_csv(\"../data/cluster/ani/cluster-ani_to_taxonomy.tsv.gz\", sep=\"\\t\", index_col=0, header=None).iloc[:,0]\n",
    "df_meta_genomes = pd.read_csv(f\"../data/cluster/mfc/{quality_label}/identifier_mapping.mfc.genomes.with_openai.tsv.gz\", sep=\"\\t\", index_col=0)\n",
    "df_meta_genome_clusters = pd.read_csv(f\"../data/cluster/mfc/{quality_label}/identifier_mapping.mfc.genome_clusters.with_openai.tsv.gz\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "df_meta_keggortholog = pd.read_csv(\"/home/ec2-user/SageMaker/s3/newatlantis-raw-veba-db-prod/VEBA/VDB_v8.1/Annotate/KOfam/kegg-ortholog_metadata.tsv\", sep=\"\\t\", index_col=0)\n",
    "keggortholog_to_name = df_meta_keggortholog[\"definition\"]\n",
    "keggmodule_to_name = pd.read_csv(\"/home/ec2-user/SageMaker/s3/newatlantis-raw-veba-db-prod/VEBA/VDB_v8.1/Annotate/KEGG-Pathway-Profiler/pathway_names.tsv.gz\", sep=\"\\t\", index_col=0, header=None).iloc[:,0]\n",
    "kegg_pathway_database = read_pickle(\"/home/ec2-user/SageMaker/s3/newatlantis-raw-veba-db-prod/VEBA/VDB_v8.1/Annotate/KEGG-Pathway-Profiler/database.pkl.gz\")\n",
    "print(\"Number of genomes: {}, Number of features: {}, Number of SLCs: {}\".format(*X_genomic_traits.shape, X_genomic_traits_clusterani.shape[0]))\n",
    "# Number of genomes: 20377, Number of features: 2124, Number of SLCs: 6719\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0811d437-ab69-4088-bf28-d41be8847db7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Clustering \n",
    "clusterer = read_pickle(f\"../data/cluster/mfc/completeness_gte90.contamination_lt5/MFC.PathwayKOfam.KNeighborsLeidenClustering.pkl\")\n",
    "clusterer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2747a9cd-ced7-483e-85f7-10f16fec09d7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# MNS\n",
    "mns = read_pickle(f\"../data/training/completeness_gte90.contamination_lt5/NAL-GDB_MNS_v2.SLC-MFC.HierarchicalNicheSpace.pkl\")\n",
    "mns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "148cd8b3-860b-4586-a605-50e0f5d44a94",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Qualitative MNS\n",
    "qualitative_mns = read_pickle(f\"../data/training/completeness_gte90.contamination_lt5/NAL-GDB_MNS_v2.SLC-MFC.QualitativeSpace.pkl\")\n",
    "qualitative_mns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cc4d8a49-df74-4ecf-a5c7-3308cab9f17e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Embedding Annnotator\n",
    "annotator = read_pickle(f\"../data/training/completeness_gte90.contamination_lt5/NAL-GDB_MNS_v2.SLC-MFC.EmbeddingAnnotator.pkl\")\n",
    "annotator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "838fadb8-d95d-4188-8a7b-58f89e0f8573",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Load KEGG module database\n",
    "\n",
    "# # Calculate KEGG module completion ratios\n",
    "# dataframes = list()\n",
    "# for id_niche, features in annotator.selected_features_.items():\n",
    "#     data = pathway_coverage_wrapper(\n",
    "#         evaluation_kos=set(features), # Annotate just the first niche\n",
    "#         database=kegg_pathway_database,\n",
    "#     )\n",
    "#     df = pd.DataFrame(data).T\n",
    "#     df.insert(0, \"name\", df.index.map(lambda x: keggmodule_to_name[x]))\n",
    "#     df.index = df.index.map(lambda x: (id_niche, x))\n",
    "#     dataframes.append(df)\n",
    "     \n",
    "# df_kegg_coverage = pd.concat(dataframes, axis=0)\n",
    "# df_kegg_coverage.to_csv(\"../data/training/completeness_gte90.contamination_lt5/NAL-GDB_MNS_v2.SLC-MFC.EmbeddingAnnotator.kegg_module_coverage.tsv.gz\", sep=\"\\t\")\n",
    "df_kegg_coverage = pd.read_csv(\"../data/training/completeness_gte90.contamination_lt5/NAL-GDB_MNS_v2.SLC-MFC.EmbeddingAnnotator.kegg_module_coverage.tsv.gz\", sep=\"\\t\", index_col=[0,1])\n",
    "\n",
    "# # Calculate KEGG module enrichment\n",
    "# dataframes = list()\n",
    "# for id_niche, features in annotator.selected_features_.items():\n",
    "#     df = unweighted_pathway_enrichment_wrapper(\n",
    "#     evaluation_kos=set(features), \n",
    "#     database=kegg_pathway_database,\n",
    "#     background_set=set(mns.X_.columns),\n",
    "#     )\n",
    "#     df.insert(0, \"name\", df.index.map(lambda x: keggmodule_to_name[x]))\n",
    "#     df.index = df.index.map(lambda x: (id_niche, x))\n",
    "#     dataframes.append(df)\n",
    "     \n",
    "# df_kegg_enrichment = pd.concat(dataframes, axis=0)\n",
    "# df_kegg_enrichment.to_csv(\"../data/training/completeness_gte90.contamination_lt5/NAL-GDB_MNS_v2.SLC-MFC.EmbeddingAnnotator.kegg_module_enrichment.tsv.gz\", sep=\"\\t\")\n",
    "df_kegg_enrichment = pd.read_csv(\"../data/training/completeness_gte90.contamination_lt5/NAL-GDB_MNS_v2.SLC-MFC.EmbeddingAnnotator.kegg_module_enrichment.tsv.gz\", sep=\"\\t\", index_col=[0,1])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d52274d6-ef31-44e1-a615-44343b269358",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Merge genomes and ANI-cluster seed observations\n",
    "df_meta_scatter = df_meta_genome_clusters.copy()\n",
    "df_meta_scatter[\"id_cluster-ani\"] = pd.NA\n",
    "df_meta_scatter = pd.concat([\n",
    "    df_meta_scatter,\n",
    "    df_meta_genomes,\n",
    "], axis=0)\n",
    "\n",
    "                \n",
    "# Sample data based on your provided DataFrame\n",
    "df_scatter = pd.concat([\n",
    "    qualitative_mns.embedding_,\n",
    "    df_meta_scatter,\n",
    "], axis=1)\n",
    "df_scatter[\"id_observation\"] = df_scatter.index\n",
    "df_scatter[\"type\"] = df_scatter[\"database\"].map(lambda x: {True:\"ANI-Cluster\", False:\"Genome\"}[pd.isnull(x)])\n",
    "\n",
    "\n",
    "# Generate 65 colors with custom lightness and saturation.\n",
    "colors = sns.hls_palette(65, l=0.6, s=0.9)\n",
    "\n",
    "# Get unique classes from your DataFrame.\n",
    "unique_classes = df_scatter[\"id_cluster-mfc\"].dropna().unique()\n",
    "\n",
    "# Create a color dictionary mapping each unique class to a hex color.\n",
    "class_to_color = {cat: mcolors.to_hex(color) for cat, color in zip(unique_classes, colors)}\n",
    "\n",
    "font_color = \"rgb(211,255,34)\"\n",
    "background_color = \"rgb(15,20,40)\"\n",
    "grid_color = \"rgb(24,30,64)\"\n",
    "\n",
    "# Create the 3D scatter plot with additional hover fields.\n",
    "fig = px.scatter_3d(\n",
    "    df_scatter,\n",
    "    x=\"PaCMAP-1\",\n",
    "    y=\"PaCMAP-2\",\n",
    "    z=\"PaCMAP-3\",\n",
    "    hover_name=\"id_observation\",\n",
    "    color=\"id_cluster-mfc\",  # color points based on this column\n",
    "    color_discrete_map=class_to_color,  # use your custom mapping\n",
    "    hover_data=[\n",
    "        \"id_cluster-ani\",\n",
    "        \"taxonomy\",\n",
    "        \"id_cluster-mfc\",\n",
    "        \"database\",\n",
    "        \"genome_size\",\n",
    "        \"domain\",\n",
    "        \"number_of_genes\",\n",
    "        \"coding_size\",\n",
    "        \"completeness\",\n",
    "        \"contamination\",\n",
    "        \"openai|gpt-4o-mini|enriched-taxa\",\n",
    "        \"openai|gpt-4o-mini|metabolism-description\",\n",
    "        \"openai|gpt-4o-mini|ecological-description\"\n",
    "    ],\n",
    "    title=qualitative_mns.name,\n",
    "\n",
    ")\n",
    "\n",
    "\n",
    "fig.update_layout(\n",
    "    height=800,\n",
    "    width=1500,\n",
    "    paper_bgcolor=background_color,  # Background color outside the plotting area\n",
    "    scene=dict(\n",
    "        bgcolor=background_color,    # 3D plot background color\n",
    "        xaxis=dict(\n",
    "            backgroundcolor=background_color,\n",
    "            gridcolor=grid_color,\n",
    "            color=font_color\n",
    "        ),\n",
    "        yaxis=dict(\n",
    "            backgroundcolor=background_color,\n",
    "            gridcolor=grid_color,\n",
    "            color=font_color\n",
    "        ),\n",
    "        zaxis=dict(\n",
    "            backgroundcolor=background_color,\n",
    "            gridcolor=grid_color,\n",
    "            color=font_color\n",
    "        )\n",
    "    ),\n",
    "    font=dict(color=font_color)\n",
    ")\n",
    "fig.update_traces(marker=dict(size=5))\n",
    "for trace in fig.data:\n",
    "    trace.marker.line = dict(color=grid_color, width=0.382)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aec273bc-bc42-433c-b231-1d1c06648072",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Merge genomes and ANI-cluster seed observations\n",
    "def query(key, df):\n",
    "    a = df[\"openai|gpt-4o-mini|metabolism-description\"].map(lambda x: key.lower() in str(x))\n",
    "    b = df[\"openai|gpt-4o-mini|ecological-description\"].map(lambda x: key.lower() in str(x))\n",
    "    return (a + b) > 0\n",
    "\n",
    "df_meta_scatter = df_meta_genome_clusters.copy()\n",
    "df_meta_scatter[\"id_cluster-ani\"] = pd.NA\n",
    "df_meta_scatter = pd.concat([\n",
    "    df_meta_scatter,\n",
    "    df_meta_genomes,\n",
    "], axis=0)\n",
    "\n",
    "                \n",
    "# Sample data based on your provided DataFrame\n",
    "df_scatter = pd.concat([\n",
    "    qualitative_mns.embedding_,\n",
    "    df_meta_scatter,\n",
    "], axis=1)\n",
    "df_scatter[\"id_observation\"] = df_scatter.index\n",
    "df_scatter[\"type\"] = df_scatter[\"database\"].map(lambda x: {True:\"ANI-Cluster\", False:\"Genome\"}[pd.isnull(x)])\n",
    "df_scatter[\"category[methane_cycle]\"] = query(\"methan\", df_scatter)\n",
    "\n",
    "\n",
    "\n",
    "# Create a color dictionary mapping each unique class to a hex color.\n",
    "class_to_color = {True:\"red\", False:\"white\"}\n",
    "\n",
    "font_color = \"rgb(211,255,34)\"\n",
    "background_color = \"rgb(15,20,40)\"\n",
    "grid_color = \"rgb(24,30,64)\"\n",
    "\n",
    "# Create the 3D scatter plot with additional hover fields.\n",
    "fig = px.scatter_3d(\n",
    "    df_scatter,\n",
    "    x=\"PaCMAP-1\",\n",
    "    y=\"PaCMAP-2\",\n",
    "    z=\"PaCMAP-3\",\n",
    "    hover_name=\"id_observation\",\n",
    "    color=\"category[methane_cycle]\",  # color points based on this column\n",
    "    color_discrete_map=class_to_color,  # use your custom mapping\n",
    "    hover_data=[\n",
    "        \"id_cluster-ani\",\n",
    "        \"taxonomy\",\n",
    "        \"id_cluster-mfc\",\n",
    "        \"database\",\n",
    "        \"genome_size\",\n",
    "        \"domain\",\n",
    "        \"number_of_genes\",\n",
    "        \"coding_size\",\n",
    "        \"completeness\",\n",
    "        \"contamination\",\n",
    "        \"openai|gpt-4o-mini|enriched-taxa\",\n",
    "        \"openai|gpt-4o-mini|metabolism-description\",\n",
    "        \"openai|gpt-4o-mini|ecological-description\"\n",
    "    ],\n",
    "    title=qualitative_mns.name,\n",
    "\n",
    ")\n",
    "\n",
    "\n",
    "fig.update_layout(\n",
    "    height=800,\n",
    "    width=1500,\n",
    "\n",
    "    paper_bgcolor=background_color,  # Background color outside the plotting area\n",
    "    scene=dict(\n",
    "        bgcolor=background_color,    # 3D plot background color\n",
    "        xaxis=dict(\n",
    "            backgroundcolor=background_color,\n",
    "            gridcolor=grid_color,\n",
    "            color=font_color\n",
    "        ),\n",
    "        yaxis=dict(\n",
    "            backgroundcolor=background_color,\n",
    "            gridcolor=grid_color,\n",
    "            color=font_color\n",
    "        ),\n",
    "        zaxis=dict(\n",
    "            backgroundcolor=background_color,\n",
    "            gridcolor=grid_color,\n",
    "            color=font_color\n",
    "        )\n",
    "    ),\n",
    "    font=dict(color=font_color)\n",
    ")\n",
    "fig.update_traces(marker=dict(size=5))\n",
    "for trace in fig.data:\n",
    "    trace.marker.line = dict(color=grid_color, width=0.382)\n",
    "fig.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eec32f18-f0c8-42bd-89ea-60d8187e6365",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "# Define your color variables\n",
    "font_color = \"rgb(211,255,34)\"\n",
    "background_color = \"rgb(15,20,40)\"\n",
    "grid_color = \"rgb(24,30,64)\"\n",
    "\n",
    "# Extract data for the query (assuming mns.diffusion_coordinates_initial_ is a DataFrame or Series)\n",
    "query = \"XB_MAG545\"\n",
    "data = mns.diffusion_coordinates_initial_.loc[query]\n",
    "\n",
    "# Convert the Series into a DataFrame:\n",
    "# The index represents the metabolic niche names and the values are the diffusion strengths.\n",
    "df_bar = data.reset_index()\n",
    "df_bar.columns = [\"Metabolic Niche\", \"Diffusion Strength\"]\n",
    "\n",
    "# Create the bar chart using Plotly Express\n",
    "fig = px.bar(df_bar, x=\"Metabolic Niche\", y=\"Diffusion Strength\")\n",
    "\n",
    "# Add a horizontal line at y=0\n",
    "fig.add_hline(y=0, line_dash=\"dash\", line_color=grid_color)\n",
    "\n",
    "# Update the layout: set the background colors and font color.\n",
    "fig.update_layout(\n",
    "    height=700,\n",
    "    width=1200,\n",
    "    paper_bgcolor=background_color,  # color outside the plot area\n",
    "    plot_bgcolor=background_color,   # color inside the plot area\n",
    "    font=dict(color=font_color),\n",
    ")\n",
    "\n",
    "# Set the bar color to be font_color\n",
    "fig.update_traces(marker_color=font_color)\n",
    "\n",
    "# Update the axis labels and tick colors\n",
    "fig.update_xaxes(title_text=\"Metabolic Niche\", color=font_color, gridcolor=grid_color)\n",
    "fig.update_yaxes(title_text=\"Diffusion Strength\", color=font_color, gridcolor=grid_color)\n",
    "\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5a2c1e5c-b47f-4769-8aaf-7a8926189860",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import plotly.express as px\n",
    "\n",
    "# Assume you have your series defined as follows:\n",
    "data = annotator.feature_weights_[\"n35\"].sort_values(ascending=False)\n",
    "data.index = data.index.map(lambda x: f\"{x} | {keggortholog_to_name[x]}\")\n",
    "\n",
    "# Convert the Series into a DataFrame\n",
    "df = data.reset_index()\n",
    "df.columns = [\"Feature\", \"Weight\"]\n",
    "\n",
    "# Create a horizontal bar chart (set orientation to \"h\")\n",
    "fig = px.bar(\n",
    "    df,\n",
    "    x=\"Weight\",\n",
    "    y=\"Feature\",\n",
    "    orientation=\"h\",\n",
    "    title=\"Feature Weights for n35\"\n",
    ")\n",
    "\n",
    "# Define your color variables\n",
    "font_color = \"rgb(211,255,34)\"\n",
    "background_color = \"rgb(15,20,40)\"\n",
    "\n",
    "# Update the layout: set the background colors and font color.\n",
    "fig.update_layout(\n",
    "    height=700,\n",
    "    width=1200,\n",
    "    paper_bgcolor=background_color,  # color outside the plot area\n",
    "    plot_bgcolor=background_color,   # color inside the plot area\n",
    "    font=dict(color=font_color),\n",
    ")\n",
    "\n",
    "# Set the bar color to be font_color\n",
    "fig.update_traces(marker_color=font_color)\n",
    "\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f557364-49c2-416f-a4ac-842732f4dc05",
   "metadata": {},
   "outputs": [],
   "source": [
    "best_iteration = pd.DataFrame(annotator.automl_results_[\"n35\"]).sort_values([\"feature_selected_testing_score\", \"feature_selected_training_score\"], ascending=[False, False]).iloc[0]\n",
    "estimator = best_iteration[\"best_estimator\"]\n",
    "# Compute actual and predicted values\n",
    "y_training = annotator.Y_[\"n35\"]\n",
    "y_training_hat = pd.Series(\n",
    "    estimator.predict(annotator.X_.loc[:, estimator.feature_names_in_]),\n",
    "    index=annotator.X_.index,\n",
    ")\n",
    "\n",
    "y_testing = annotator.Y_testing_[\"n35\"]\n",
    "y_testing_hat = pd.Series(\n",
    "    estimator.predict(annotator.X_testing_.loc[:, estimator.feature_names_in_]),\n",
    "    index=annotator.X_testing_.index,\n",
    ")\n",
    "\n",
    "# Combine the training and testing data\n",
    "y = pd.concat([y_training, y_testing])\n",
    "y_hat = pd.concat([y_training_hat, y_testing_hat])\n",
    "\n",
    "# Create a DataFrame for plotting. We label points as \"Training\" or \"Testing\".\n",
    "df_scatter = pd.DataFrame({\n",
    "    \"Actual\": y,\n",
    "    \"Predicted\": y_hat,\n",
    "    \"Group\": [\"Training\"] * y_training.size + [\"Testing\"] * y_testing.size\n",
    "})\n",
    "\n",
    "# Create a scatter plot with custom colors for each group.\n",
    "fig = px.scatter(\n",
    "    df_scatter,\n",
    "    x=\"Actual\",\n",
    "    y=\"Predicted\",\n",
    "    color=\"Group\",\n",
    "    color_discrete_map={\"Training\": \"white\", \"Testing\": font_color},\n",
    "    title=\"Actual vs Predicted for n35\"\n",
    ")\n",
    "\n",
    "# Define your color scheme\n",
    "font_color = \"rgb(211,255,34)\"\n",
    "background_color = \"rgb(15,20,40)\"\n",
    "grid_color = \"rgb(24,30,64)\"\n",
    "\n",
    "# Update the layout with your color theme\n",
    "fig.update_layout(\n",
    "    height=700,\n",
    "    width=1200,\n",
    "    paper_bgcolor=background_color,\n",
    "    plot_bgcolor=background_color,\n",
    "    font=dict(color=font_color)\n",
    ")\n",
    "fig.update_xaxes(gridcolor=grid_color, color=font_color)\n",
    "fig.update_yaxes(gridcolor=grid_color, color=font_color)\n",
    "fig.update_traces(marker=dict(size=5))\n",
    "for trace in fig.data:\n",
    "    trace.marker.line = dict(color=grid_color, width=0.382)\n",
    "fig.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5885e982-b39c-4115-a923-7046e08a1568",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "list(map(lambda x: (x, keggortholog_to_name[x]), kegg_pathway_database[\"M00175\"][\"ko_to_nodes\"].keys()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1e06e344-48d9-4c3b-b80b-3fc97c586a57",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load credentials\n",
    "config = dotenv_values(\"/home/ec2-user/SageMaker/.openai\")\n",
    "\n",
    "# Setup client\n",
    "llm = LLMAnnotator(**config, description=\"EmbeddingAnnotator Contextualizer\")\n",
    "llm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad1587fd-150e-484a-bee8-9f5ab09fdd9a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "prompts = dict()\n",
    "prompts[\"annotate_embeddings[kegg_orthogs=True,genome_taxonomy=False]\"] = \"\"\"\n",
    "You are an AI assistant tasked with analyzing a dictionary of KEGG orthologs (features) and their corresponding weights. \n",
    "These weights indicate the predictive capacity of each ortholog for a specific dimension in a diffusion map embedding. \n",
    "Your goal is to assign high-level metabolic or ecological annotations to this embedding dimension, providing insight into\n",
    "the metabolic niche of organisms with high magnitude along this dimension.\n",
    "\n",
    "Here is the dictionary of features (KEGG orthologs) and their weights:\n",
    "<feature_weight_dict>\n",
    "%s\n",
    "</feature_weight_dict>\n",
    "\n",
    "To complete this task, follow these steps:\n",
    "1. Examine the features and weights in the dictionary. Pay attention to the magnitude and sign of the weights. \n",
    "Higher magnitude will mean these proteins will be the most influential in predicting the diffusion coordinate\n",
    "while the sign indicates a separation along some data-driven axis likely representing complex metabolic patterns.\n",
    "2. Use your knowledge of KEGG pathways and biological processes to understand the function of these orthologs \n",
    "and their roles in metabolic pathways.\n",
    "3. Contextualize your research by searching for common themes or related processes among the weighted features.\n",
    "4. Based on your analysis, formulate a high-level metabolic or ecological annotation for this embedding dimension. \n",
    "Consider what type of organisms or metabolic processes might be associated with high magnitude along this dimension.\n",
    "Provide your answer in json format that be read into a Python dictionary without additional parsing:\n",
    "{\n",
    "\"summary\":[Here, provide a consise analysis of your findings in and how the features relate to each other metabolically and ecologically.],\n",
    "\"metabolic_context\":[Provide a concise, high-level annotation of the metabolic context for the embedding dimension based on your analysis.],\n",
    "\"ecological_context\":[Provide a concise, high-level annotation of the ecological context for the embedding dimension based on your analysis.]\n",
    "\"justification\":[Explain your reasoning for the annotation, citing specific features and their weights from the dictionary. Discuss any potential limitations or uncertainties in your interpretation.]\n",
    "}\n",
    "Remember to think critically about the biological significance of these features and their weights. Consider how they might work together in metabolic pathways or ecological processes. \n",
    "If you're unsure about certain aspects, acknowledge this uncertainty in your justification.\n",
    "Your goal is to provide a well-reasoned, biologically meaningful annotation that could help researchers understand the metabolic or ecological significance of this embedding dimension.\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6c85a7b7-87f5-4357-9da2-365adedab625",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "info = annotator.feature_weights_[\"n35\"].sort_values(ascending=False)\n",
    "info.index = info.index.map(lambda x: (x,keggortholog_to_name[x]))\n",
    "info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9846b795-da22-45fb-be9c-65fb588a43ad",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "prompt = prompts[\"annotate_embeddings[kegg_orthogs=True,genome_taxonomy=False]\"]%(str(info.to_dict()))\n",
    "response = llm.query(prompt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c4de3c0-aa40-4ce9-b624-72372427a9b1",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "d = json.load(StringIO(response))\n",
    "for k, v in d.items():\n",
    "    print(format_header(k))\n",
    "    print(v)\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3b844b20-7181-4104-bd00-ba73522bcf06",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77ea55e0-95e2-4def-b1cd-e9bfc0f31f4b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_mns",
   "language": "python",
   "name": "conda_mns"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
