{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3fb3da35-c420-48a5-bfb3-852785159a74",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/ec2-user/SageMaker/environments/mns/lib/python3.9/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    }
   ],
   "source": [
    "# -*- coding: utf-8 -*-\n",
    "import os\n",
    "import sys\n",
    "from collections import defaultdict\n",
    "from tqdm import tqdm\n",
    "import numpy as np # Can't install NumPy 2.2.2 which is what the pkls were saved with\n",
    "import pandas as pd # 'v2.2.3'\n",
    "# import anndata as ad\n",
    "from scipy.spatial.distance import (\n",
    "    pdist,\n",
    "    squareform,\n",
    ")\n",
    "from pyexeggutor import (\n",
    "    build_logger,\n",
    "    write_pickle,\n",
    "    read_pickle,\n",
    "    check_argument_choice,\n",
    "    format_header,\n",
    ")\n",
    "\n",
    "from sklearn.metrics import (\n",
    "    pairwise_distances,\n",
    "    silhouette_score,\n",
    ")\n",
    "import ensemble_networkx as enx\n",
    "import igraph as ig\n",
    "import optuna\n",
    "import joblib\n",
    "\n",
    "# Metabolic Niche Space\n",
    "from metabolic_niche_space.neighbors import (\n",
    "    convert_distance_matrix_to_kneighbors_matrix,\n",
    ")\n",
    "from metabolic_niche_space.utils import (\n",
    "    compile_parameter_space,\n",
    ")\n",
    "\n",
    "\n",
    "# Metabolic Niche Space\n",
    "from metabolic_niche_space.neighbors import (\n",
    "    KNeighborsLeidenClustering,\n",
    "    pairwise_distances_kneighbors,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "121602ba-029e-4622-a30c-3b1e040a3776",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of genomes: 20377, Number of features: 2124, Number of SLCs: 6719\n",
      "CPU times: user 22.6 s, sys: 1.51 s, total: 24.1 s\n",
      "Wall time: 25.5 s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "# Data\n",
    "df_quality = pd.read_csv(\"../data/quality.tsv.gz\",sep=\"\\t\", index_col=0)\n",
    "\n",
    "quality_label=\"completeness_gte90.contamination_lt5\"\n",
    "# quality_label=\"completeness_gte50.contamination_lt10\"\n",
    "output_directory=f\"../data/cluster/mfc/{quality_label}\"\n",
    "os.makedirs(output_directory, exist_ok=True)\n",
    "\n",
    "genome_to_clusterani = pd.read_csv(f\"../data/training/{quality_label}/y.tsv.gz\", sep=\"\\t\", index_col=0, header=None).iloc[:,0].astype(\"category\")\n",
    "X_genomic_traits = pd.read_csv(f\"../data/training/{quality_label}/X.tsv.gz\", sep=\"\\t\", index_col=0).astype(bool)\n",
    "X_genomic_traits_clusterani = pd.read_csv(f\"../data/training/{quality_label}/X_grouped.tsv.gz\", sep=\"\\t\", index_col=0).astype(bool)\n",
    "# eukaryotes = read_list(f\"../data/cluster/ani/eukaryotic/{quality_label}/organisms.list\", set)\n",
    "# prokaryotes = read_list(f\"../data/cluster/ani/prokaryotic/{quality_label}/organisms.list\", set)\n",
    "\n",
    "genome_to_taxonomy = pd.read_csv(\"../data/taxonomy.tsv.gz\", sep=\"\\t\", index_col=0).iloc[:,0]\n",
    "clusterani_to_taxonomy = pd.read_csv(\"../data/cluster/ani/cluster-ani_to_taxonomy.tsv.gz\", sep=\"\\t\", index_col=0, header=None).iloc[:,0]\n",
    "df_meta_mfc__genomes = pd.read_csv(f\"../data/cluster/mfc/{quality_label}/identifier_mapping.mfc.genomes.with_openai.tsv.gz\", sep=\"\\t\", index_col=0)\n",
    "df_meta_mfc__slc = pd.read_csv(f\"../data/cluster/mfc/{quality_label}/identifier_mapping.mfc.genome_clusters.with_openai.tsv.gz\", sep=\"\\t\", index_col=0)\n",
    "\n",
    "X_genomic_traits_mfc = X_genomic_traits_clusterani.groupby(df_meta_mfc__slc[\"id_cluster-mfc\"]).sum() > 0\n",
    "df_kegg = pd.read_csv(\"/home/ec2-user/SageMaker/s3/newatlantis-raw-veba-db-prod/VDB_v8.1/Annotate/KOfam/kegg-ortholog_metadata.tsv\", sep=\"\\t\", index_col=0)\n",
    "ko_to_description = df_kegg[\"definition\"]\n",
    "\n",
    "print(\"Number of genomes: {}, Number of features: {}, Number of SLCs: {}\".format(*X_genomic_traits.shape, X_genomic_traits_clusterani.shape[0]))\n",
    "# Number of genomes: 20377, Number of features: 2124, Number of SLCs: 6719\n",
    "\n",
    "# CPU times: user 3.41 s, sys: 87.8 ms, total: 3.5 s\n",
    "# Wall time: 3.5 s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "19421d59-ebb9-4493-933b-9f7a33d54cb1",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "8b5cb7d4-e2a8-4579-938f-72414db9c600",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def is_square_symmetric(matrix, tol=1e-8):\n",
    "    \"\"\"Check if a matrix is square and symmetric.\"\"\"\n",
    "    matrix = np.array(matrix)  # Ensure it's a NumPy array\n",
    "    return matrix.shape[0] == matrix.shape[1] and np.allclose(matrix, matrix.T, atol=tol)\n",
    "\n",
    "def stop_when_exceeding_trials(n_trials, logger):\n",
    "    def callback(study, trial):\n",
    "        \"\"\"\n",
    "        Callback that stops optimization if the total number of trials exceeds `n_trials`.\n",
    "        \"\"\"\n",
    "        finished_trial_states = [\n",
    "            optuna.trial.TrialState.COMPLETE, \n",
    "            optuna.trial.TrialState.FAIL, \n",
    "            optuna.trial.TrialState.RUNNING,\n",
    "        ]\n",
    "\n",
    "        total_completed_trials = sum(1 for t in study.trials if t.state in finished_trial_states)\n",
    "\n",
    "        if total_completed_trials >= n_trials:\n",
    "            logger.warning(f\"[Callback] Stopping optimization: {total_completed_trials} trials reached (limit={n_trials})\")\n",
    "            study.stop()\n",
    "    \n",
    "    return callback  # Return the function with access to `n_trials` and `logger`\n",
    "                  \n",
    "\n",
    "class KNeighborsLeidenClustering(object):\n",
    "    def __init__(\n",
    "        self, \n",
    "        # General\n",
    "        name:str=None,\n",
    "        observation_type:str=None,\n",
    "        feature_type:str=None,\n",
    "        class_type:str=None,\n",
    "        method:str = \"one_minus\",\n",
    "        initial_distance_metric:str=\"precomputed\",\n",
    "        scoring_distance_metric:str=\"euclidean\",\n",
    "        n_neighbors:int=\"auto\",\n",
    "        \n",
    "        # Community detection\n",
    "        n_iter=10, \n",
    "        converge_iter=-1,\n",
    "        minimum_membership_consistency=1.0, \n",
    "        cluster_prefix=\"c\",\n",
    "        \n",
    "        # Optuna\n",
    "        n_trials=10,\n",
    "        n_jobs:int=1,\n",
    "        n_concurrent_trials:int=1,\n",
    "        initial_params:dict = None,\n",
    "        objective_direction=\"maximize\",\n",
    "        checkpoint_directory=None,\n",
    "        study_timeout=None,\n",
    "        study_callbacks=None,\n",
    "        random_state=0,\n",
    "        verbose=1,\n",
    "        stream=sys.stdout,\n",
    "        ):\n",
    "        \n",
    "        # General\n",
    "        if name is None:\n",
    "            name = str(uuid.uuid4())\n",
    "            \n",
    "        self.name = name\n",
    "        self.observation_type = observation_type\n",
    "        self.feature_type = feature_type\n",
    "        self.class_type = class_type        \n",
    "        self.method = method\n",
    "        self.initial_distance_metric = initial_distance_metric\n",
    "        self.scoring_distance_metric = scoring_distance_metric\n",
    "        \n",
    "        # Community detection\n",
    "        self.n_iter=n_iter\n",
    "        self.converge_iter=converge_iter\n",
    "        self.minimum_membership_consistency=minimum_membership_consistency\n",
    "        self.cluster_prefix = cluster_prefix\n",
    "        \n",
    "        # Optuna\n",
    "        self.n_jobs = n_jobs\n",
    "        self.n_trials = n_trials\n",
    "        self.n_concurrent_trials = n_concurrent_trials\n",
    "        self.initial_params = initial_params\n",
    "        self.checkpoint_directory = checkpoint_directory\n",
    "        self.random_state = random_state\n",
    "        self.study_timeout = study_timeout\n",
    "        if study_callbacks is None:\n",
    "            study_callbacks = []\n",
    "        self.study_callbacks = study_callbacks\n",
    "        self.objective_direction = objective_direction\n",
    "\n",
    "        # Hyperparameters\n",
    "        self.is_tuned = True\n",
    "        if n_neighbors == \"auto\":\n",
    "            n_neighbors = [int, 10, 100]\n",
    "            self.is_tuned = False\n",
    "        if isinstance(n_neighbors, list):\n",
    "            self.is_tuned = False\n",
    "        self.n_neighbors = n_neighbors\n",
    "        \n",
    "        self._param_space = dict(\n",
    "            n_neighbors = self.n_neighbors,\n",
    "        )\n",
    "        \n",
    "        self.logger = build_logger(self.name, stream=stream)\n",
    "        self.verbose = verbose\n",
    "        self.is_fitted = False\n",
    "        \n",
    "    @staticmethod\n",
    "    def one_minus(distance_matrix):\n",
    "        \"\"\"1 - distance transformation (e.g., Jaccard similarity).\"\"\"\n",
    "        return 1 - distance_matrix\n",
    "\n",
    "    @staticmethod\n",
    "    def inverse(distance_matrix):\n",
    "        \"\"\"1 / (1 + distance) transformation to avoid division by zero.\"\"\"\n",
    "        return 1 / (1 + distance_matrix)\n",
    "\n",
    "    @staticmethod\n",
    "    def exponential(distance_matrix, sigma=1.0):\n",
    "        \"\"\"Exponential decay transformation (e.g., Gaussian Kernel).\"\"\"\n",
    "        return np.exp(-distance_matrix / sigma)\n",
    "        \n",
    "    def distance_to_similarity(self, distances):\n",
    "        \"\"\"Applies the selected transformation method.\"\"\"\n",
    "        transformations = {\n",
    "            \"one_minus\": self.one_minus,\n",
    "            \"inverse\": self.inverse,\n",
    "            \"exponential\": self.exponential,\n",
    "        }\n",
    "\n",
    "        if self.method not in transformations:\n",
    "            raise ValueError(f\"Unknown method: {self.method}. Choose from {list(transformations.keys())}\")\n",
    "\n",
    "        return transformations[self.method](distances)\n",
    "    \n",
    "    def tune(\n",
    "        self,\n",
    "        distance_matrix:pd.DataFrame,\n",
    "        sampler, \n",
    "        **study_kws,\n",
    "        ):\n",
    "\n",
    "        def _objective(trial):\n",
    "            try:\n",
    "\n",
    "                # Compile parameters\n",
    "                params = compile_parameter_space(\n",
    "                    trial, \n",
    "                    self._param_space,\n",
    "                )\n",
    "\n",
    "                # Parameters\n",
    "                n_neighbors = params[\"n_neighbors\"]\n",
    "\n",
    "                if n_neighbors >= distance_matrix.shape[0]:\n",
    "                    raise ValueError(f\"n_neighbors {n_neighbors} is larger than the number of observations {distance_matrix.shape[0]}\")\n",
    "                else:\n",
    "\n",
    "                    # Convert distance matrix to non-redundant KNN\n",
    "                    if self.verbose > 1: self.logger.info(f\"[Trial {trial.number}] Convert distance matrix to non-redundant KNN, remove disconnected nodes, and convert to similarity graph: n_neighbors={n_neighbors}\")\n",
    "                    knn = convert_distance_matrix_to_kneighbors_matrix(distance_matrix, n_neighbors=n_neighbors, redundant_form=False)\n",
    "\n",
    "                    # Remove disconnected nodes and convert to similarity\n",
    "                    knn = knn[knn > 0]\n",
    "                    knn_similarity = self.distance_to_similarity(knn)\n",
    "                    del knn\n",
    "\n",
    "                    # Convert KNN to iGraph\n",
    "                    graph = enx.convert_network(knn_similarity, ig.Graph)\n",
    "\n",
    "                    # Identify leiden communities with multiple seeds\n",
    "                    if self.verbose > 1: self.logger.info(f\"[Trial {trial.number}] Identify Leiden communities, membership co-occurrence ratios, and building clustered graph: n_neighbors={n_neighbors}\")\n",
    "                    progressbar_message = None\n",
    "                    df_communities = enx.community_detection(graph, n_iter=self.n_iter, converge_iter=self.converge_iter, n_jobs=1, progressbar_message=progressbar_message)\n",
    "\n",
    "                    # Identify membership co-occurrence ratios\n",
    "                    node_pair_membership_cooccurrences = enx.community_membership_cooccurrence(df_communities).mean(axis=1)\n",
    "                    del df_communities\n",
    "\n",
    "                    # Identify node pairs that have co-membership 100% of the time\n",
    "                    node_pairs_with_consistent_membership = set(node_pair_membership_cooccurrences[lambda x: x >= self.minimum_membership_consistency].index)\n",
    "                    del node_pair_membership_cooccurrences\n",
    "\n",
    "                    # Get list of clustered edges\n",
    "                    clustered_edgelist = enx.get_undirected_igraph_edgelist_indices(graph, node_pairs_with_consistent_membership)\n",
    "                    del node_pairs_with_consistent_membership\n",
    "\n",
    "                    # Get clustered graph\n",
    "                    graph_clustered = graph.subgraph_edges(clustered_edgelist, delete_vertices=True)\n",
    "                    node_to_cluster = pd.Series(enx.get_undirected_igraph_connected_components(graph_clustered))\n",
    "                    del graph\n",
    "\n",
    "                    # Calculate silhouette scores\n",
    "                    if self.verbose > 1: self.logger.info(f\"[Trial {trial.number}] Calculating silhouette scores: n_neighbors={n_neighbors}\")\n",
    "                    clustered_nodes = node_to_cluster.index\n",
    "                    index = clustered_nodes.map(lambda x: distance_matrix.index.get_loc(x)).values\n",
    "                    dist = distance_matrix.values[index,:][:,index]\n",
    "                    score = silhouette_score(dist, node_to_cluster.values, metric=\"precomputed\", sample_size=None, random_state=None) \n",
    "                    del dist\n",
    "\n",
    "                    return score\n",
    "\n",
    "            except Exception as e:\n",
    "                self.logger.error(f\"[Trial {trial.number}] Failed due to error: {e}. Marking as pruned.\")\n",
    "                raise optuna.TrialPruned()  # Prevents skipping trials\n",
    "\n",
    "            finally:\n",
    "                if self.checkpoint_directory:\n",
    "                    joblib.dump(study, os.path.join(self.checkpoint_directory, f\"{self.name}.pkl\"))  # Save checkpoint\n",
    "\n",
    "        # Sampler\n",
    "        if sampler is None:\n",
    "            sampler = optuna.samplers.TPESampler(seed=self.random_state)\n",
    "\n",
    "        # Study\n",
    "        study_params = {\n",
    "            \"direction\":self.objective_direction, \n",
    "            \"study_name\":self.name, \n",
    "            \"sampler\":sampler, \n",
    "            **study_kws,\n",
    "        }\n",
    "\n",
    "        # Checkpoints\n",
    "        study = None\n",
    "        if self.checkpoint_directory:\n",
    "            if not os.path.exists(self.checkpoint_directory):\n",
    "                if self.verbose > 1: self.logger.info(f\"Creating checkpoint directory: {self.checkpoint_directory}\")\n",
    "                os.makedirs(self.checkpoint_directory)\n",
    "\n",
    "            serialized_checkpoint_filepath = os.path.join(self.checkpoint_directory, f\"{self.name}.pkl\")\n",
    "\n",
    "            if os.path.exists(serialized_checkpoint_filepath):\n",
    "                if self.verbose > 1: self.logger.info(f\"[Loading] Checkpoint file: {serialized_checkpoint_filepath}\")\n",
    "                study = joblib.load(serialized_checkpoint_filepath)\n",
    "            else:\n",
    "                if self.verbose > 1: self.logger.info(f\"[Creating] Checkpoint file: {serialized_checkpoint_filepath}\")\n",
    "\n",
    "        if study is None:\n",
    "            study = optuna.create_study(**study_params)\n",
    "\n",
    "        if self.initial_params:\n",
    "            if self.verbose > 1: self.logger.info(f\"Adding initial parameters to study: {self.initial_params}\")\n",
    "            study.enqueue_trial(self.initial_params, user_attrs={\"memo\": \"initial_params\"}, skip_if_exists=True)\n",
    "\n",
    "        # Optimize\n",
    "        callback_fn = stop_when_exceeding_trials(self.n_trials, self.logger)\n",
    "\n",
    "        study.optimize(\n",
    "            _objective, \n",
    "            n_trials=self.n_trials, \n",
    "            n_jobs=self.n_concurrent_trials,\n",
    "            timeout=self.study_timeout, \n",
    "            show_progress_bar=self.verbose >= 2, \n",
    "            callbacks=self.study_callbacks + [callback_fn], \n",
    "            gc_after_trial=True,\n",
    "        )\n",
    "\n",
    "        return study\n",
    "\n",
    "    def fit(\n",
    "        self,\n",
    "        X:pd.DataFrame,\n",
    "        sampler=None,\n",
    "        copy=\"auto\",\n",
    "        **study_kws,\n",
    "        ):\n",
    "\n",
    "        # Check inputs\n",
    "        if self.initial_distance_metric == \"precomputed\":\n",
    "            if len(X.shape) == 1:\n",
    "                X = squareform(X)\n",
    "            if not is_square_symmetric(X):\n",
    "                raise ValueError(\"If X is precomputed, it must be square and symmetric\")\n",
    "            distance_matrix = X\n",
    "        else:\n",
    "            if self.initial_distance_metric == \"jaccard\":\n",
    "                X = X.astype(bool)\n",
    "            if self.verbose > 0:\n",
    "                self.logger.info(\"[Start] Processing distance matrix\")\n",
    "            if self.initial_distance_metric == \"euclidean\":\n",
    "                distance_matrix = squareform(pdist(X.values, metric=self.initial_distance_metric))\n",
    "            else:\n",
    "                distance_matrix = pairwise_distances(X=X.values, metric=self.initial_distance_metric, n_jobs=self.n_jobs)\n",
    "\n",
    "        if self.verbose > 0:\n",
    "            self.logger.info(\"[End] Processing distance matrix\")\n",
    "\n",
    "        # Store\n",
    "        if copy == \"auto\":\n",
    "            if self.initial_distance_metric == \"precomputed\":\n",
    "                copy = False\n",
    "            else:\n",
    "                copy = True\n",
    "        if copy:\n",
    "            self.X_ = X.copy()\n",
    "\n",
    "        # Tune\n",
    "        if not self.is_tuned:\n",
    "            if self.verbose > 0:\n",
    "                self.logger.info(\"[Begin] Hyperparameter Tuning\")\n",
    "            self.study_ = self.tune(\n",
    "                distance_matrix=distance_matrix,\n",
    "                sampler=sampler, \n",
    "                **study_kws,\n",
    "                )\n",
    "            for k, v in self.study_.best_params.items():\n",
    "                setattr(self,k,v)\n",
    "            if self.verbose > 0:\n",
    "                self.logger.info(f\"Tuned parameters (Score={self.study_.best_value}): {self.study_.best_params}\")\n",
    "                self.logger.info(\"[End] Hyperparameter Tuning\")\n",
    "            self.is_tuned = True\n",
    "\n",
    "        # Convert distance matrix to non-redundant KNN\n",
    "        knn = convert_distance_matrix_to_kneighbors_matrix(distance_matrix, n_neighbors=self.n_neighbors, redundant_form=False)\n",
    "\n",
    "        # Remove disconnected nodes and convert to similarity\n",
    "        knn = knn[knn > 0]\n",
    "        knn_similarity = self.distance_to_similarity(knn)\n",
    "        del knn\n",
    "\n",
    "        # Convert KNN to iGraph\n",
    "        graph = enx.convert_network(knn_similarity, ig.Graph)\n",
    "\n",
    "        # Identify leiden communities with multiple seeds\n",
    "        df_communities = enx.community_detection(graph, n_iter=self.n_iter, converge_iter=self.converge_iter, n_jobs=1)\n",
    "\n",
    "        # Identify membership co-occurrence ratios\n",
    "        node_pair_membership_cooccurrences = enx.community_membership_cooccurrence(df_communities).mean(axis=1)\n",
    "        del df_communities\n",
    "\n",
    "        # Identify node pairs that have co-membership 100% of the time\n",
    "        node_pairs_with_consistent_membership = set(node_pair_membership_cooccurrences[lambda x: x >= self.minimum_membership_consistency].index)\n",
    "        del node_pair_membership_cooccurrences\n",
    "\n",
    "        # Get list of clustered edges\n",
    "        clustered_edgelist = enx.get_undirected_igraph_edgelist_indices(graph, node_pairs_with_consistent_membership)\n",
    "        del node_pairs_with_consistent_membership\n",
    "\n",
    "        # Get clustered graph\n",
    "        self.graph_clustered_ = graph.subgraph_edges(clustered_edgelist, delete_vertices=True)\n",
    "        self.labels_ = pd.Series(enx.get_undirected_igraph_connected_components(self.graph_clustered_, cluster_prefix=self.cluster_prefix))\n",
    "        del graph\n",
    "\n",
    "        # Calculate silhouette scores\n",
    "        clustered_nodes = self.labels_.index\n",
    "        index = clustered_nodes.map(lambda x: distance_matrix.index.get_loc(x)).values\n",
    "        dist = distance_matrix.values[index,:][:,index]\n",
    "        self.score_ = silhouette_score(dist, self.labels_.values, metric=\"precomputed\", sample_size=None, random_state=None) \n",
    "        del dist\n",
    "        \n",
    "        self.n_observations_ = distance_matrix.shape[0]\n",
    "        self.n_clusters_ = self.labels_.nunique()\n",
    "\n",
    "        self.is_fitted = True\n",
    "\n",
    "        return self\n",
    "    \n",
    "    def fit_transform(\n",
    "        self,\n",
    "        **kws,\n",
    "        ):\n",
    "\n",
    "        self.fit(**kws)\n",
    "        return self.labels_\n",
    "            \n",
    "    def to_file(self, filepath):\n",
    "        write_pickle(self, filepath)\n",
    "\n",
    "    # =======\n",
    "    # Built-in\n",
    "    # =======\n",
    "    def __repr__(self):\n",
    "        pad = 4\n",
    "        header = format_header(f\"{self.__class__.__name__}(Name:{self.name}, ObservationType: {self.observation_type}, FeatureType: {self.feature_type})\", line_character=\"=\")\n",
    "\n",
    "        n = len(header.split(\"\\n\")[0])\n",
    "        fields = [\n",
    "            header,\n",
    "            pad*\" \" + \"* initial_distance_metric: {}\".format(self.initial_distance_metric),\n",
    "            pad*\" \" + \"* scoring_distance_metric: {}\".format(self.scoring_distance_metric),\n",
    "            pad*\" \" + \"* cluster_prefix: {}\".format(self.cluster_prefix),\n",
    "            pad*\" \" + \"* checkpoint_directory: {}\".format(self.checkpoint_directory),\n",
    "        ]\n",
    "                                                  \n",
    "        if self.is_tuned:\n",
    "            fields += [\n",
    "            pad*\" \" + \"* n_neighbors: {}\".format(self.n_neighbors),\n",
    "            pad*\" \" + \"* score: {}\".format(self.score_),\n",
    "            ]\n",
    "        if self.is_fitted:\n",
    "            fields += [\n",
    "            pad*\" \" + \"* n_observations: {}\".format(self.n_observations_),\n",
    "            pad*\" \" + \"* n_clusters: {}\".format(self.n_clusters_),\n",
    "            ]  \n",
    "\n",
    "        return \"\\n\".join(fields)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "240fd44a-0496-4d5d-a688-3cb764a2e2aa",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "jaccard_distances = read_pickle(f\"{output_directory}/genomic_traits_clusterani.jaccard_distance.dataframe.pkl\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a74fe5d4-359c-49ea-aa6e-969930cdda62",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "n = 250\n",
    "D = jaccard_distances.iloc[:n,:n]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "a8598732-daf7-4abc-b816-f2666a64f8df",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-02-17 23:20:17,483 - test-clustering - INFO - [End] Processing distance matrix\n",
      "2025-02-17 23:20:17,484 - test-clustering - INFO - [Begin] Hyperparameter Tuning\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[I 2025-02-17 23:20:17,484] A new study created in memory with name: test-clustering\n",
      "[I 2025-02-17 23:20:26,532] Trial 0 finished with value: -0.034778178453893825 and parameters: {'n_neighbors': 59}. Best is trial 0 with value: -0.034778178453893825.\n",
      "[I 2025-02-17 23:20:41,971] Trial 1 finished with value: 0.18770045861209358 and parameters: {'n_neighbors': 75}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:20:56,359] Trial 2 finished with value: -0.09340670765810676 and parameters: {'n_neighbors': 64}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:21:09,724] Trial 3 finished with value: -0.034778178453893825 and parameters: {'n_neighbors': 59}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:21:18,851] Trial 4 finished with value: -0.0035042243356014496 and parameters: {'n_neighbors': 48}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:21:32,768] Trial 5 finished with value: -0.10878585335693937 and parameters: {'n_neighbors': 68}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:21:42,208] Trial 6 finished with value: -0.002683175800124766 and parameters: {'n_neighbors': 49}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:22:01,240] Trial 7 finished with value: 0.17939654211940825 and parameters: {'n_neighbors': 91}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:22:21,730] Trial 8 finished with value: 0.17836311503405655 and parameters: {'n_neighbors': 97}. Best is trial 1 with value: 0.18770045861209358.\n",
      "[I 2025-02-17 23:22:28,849] Trial 9 finished with value: 0.019881539910157432 and parameters: {'n_neighbors': 44}. Best is trial 1 with value: 0.18770045861209358.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2025-02-17 23:22:31,798 - test-clustering - WARNING - [Callback] Stopping optimization: 10 trials reached (limit=10)\n",
      "2025-02-17 23:22:31,799 - test-clustering - INFO - Tuned parameters (Score=0.18770045861209358): {'n_neighbors': 75}\n",
      "2025-02-17 23:22:31,800 - test-clustering - INFO - [End] Hyperparameter Tuning\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Community detection: 100%|██████████| 10/10 [00:03<00:00,  3.21it/s]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "==========================================================================================\n",
       "KNeighborsLeidenClustering(Name:test-clustering, ObservationType: None, FeatureType: None)\n",
       "==========================================================================================\n",
       "    * initial_distance_metric: precomputed\n",
       "    * scoring_distance_metric: euclidean\n",
       "    * cluster_prefix: c\n",
       "    * checkpoint_directory: None\n",
       "    * n_neighbors: 75\n",
       "    * score: 0.18770045861209358\n",
       "    * n_observations: 250\n",
       "    * n_clusters: 2"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster = KNeighborsLeidenClustering(\n",
    "    name=\"test-clustering\",\n",
    "    initial_distance_metric=\"precomputed\", \n",
    "    n_trials=10,\n",
    ")\n",
    "cluster.fit(D)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "3462ef3a-2c6a-4898-82dc-f15f86332c39",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "==========================================================================================\n",
       "KNeighborsLeidenClustering(Name:test-clustering, ObservationType: None, FeatureType: None)\n",
       "==========================================================================================\n",
       "    * initial_distance_metric: precomputed\n",
       "    * scoring_distance_metric: euclidean\n",
       "    * cluster_prefix: c\n",
       "    * checkpoint_directory: checkpoints\n",
       "    * n_neighbors: 48\n",
       "    * score: -0.0035042243356014496\n",
       "    * n_observations: 250\n",
       "    * n_clusters: 250"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cluster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "51ad81fb-1f0c-4d9e-9a87-74353d6abf19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "649a9ca0-08c7-4a8b-a9de-5108d2e4661e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8e120cbf-62d7-4c3b-98f8-b3954f28ff7d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# jaccard_distances = read_pickle(f\"{output_directory}/genomic_traits_clusterani.jaccard_distance.dataframe.pkl\")\n",
    "\n",
    "\n",
    "\n",
    "# study_name = \"MetabolicFunctionalClass Detection\"\n",
    "# logger = build_logger(study_name)\n",
    "\n",
    "# for n_neighbors in [50,60,70,80][::-1]:\n",
    "#     n = jaccard_distances.shape[0]\n",
    "    \n",
    "#     # low = int(np.log(n))\n",
    "#     # high = int(np.sqrt(n))\n",
    "\n",
    "#     # Parameters to tune (This section either grabs redundant trials or gets stuck in a while loop\n",
    "#     #while True:\n",
    "#     # n_neighbors = trial.suggest_int(\"n_neighbors\", low, high)\n",
    "    \n",
    "#     # Convert distance matrix to non-redundant KNN\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Convert distance matrix to non-redundant KNN\")\n",
    "#     knn = convert_distance_matrix_to_kneighbors_matrix(jaccard_distances, n_neighbors=n_neighbors, redundant_form=False)\n",
    "    \n",
    "#     # Remove disconnected nodes and convert to similarity\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Remove disconnected nodes and convert to similarity\")\n",
    "#     knn_similarity = 1 - knn[knn > 0]\n",
    "#     del knn\n",
    "\n",
    "#     # Convert KNN to iGraph\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Convert KNN to iGraph (n_neighbors = {n_neighbors})\")\n",
    "#     graph = enx.convert_network(knn_similarity, ig.Graph)\n",
    "\n",
    "#     # Identify leiden communities with multiple seeds\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Identify leiden communities with multiple seeds\")\n",
    "#     progressbar_message = f\"[n_neighbors = {n_neighbors}] Community detection\"\n",
    "#     df_communities = enx.community_detection(graph, n_iter=100, converge_iter=-1, n_jobs=1, progressbar_message=progressbar_message)\n",
    "    \n",
    "#     # Identify membership co-occurrence ratios\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Identify membership co-occurrence ratios\")\n",
    "#     node_pair_membership_cooccurrences = enx.community_membership_cooccurrence(df_communities).mean(axis=1)\n",
    "    \n",
    "#     # Identify node pairs that have co-membership 100% of the time\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Identify node pairs that have co-membership 100% of the time\")\n",
    "\n",
    "#     node_pairs_with_consistent_membership = set(node_pair_membership_cooccurrences[lambda x: x == 1.0].index)\n",
    "\n",
    "#     # Get list of clustered edges\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Get list of clustered edges\")\n",
    "\n",
    "#     clustered_edgelist = enx.get_undirected_igraph_edgelist_indices(graph, node_pairs_with_consistent_membership)\n",
    "    \n",
    "#     # Get clustered graph\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Get clustered graph\")\n",
    "\n",
    "#     graph_clustered = graph.subgraph_edges(clustered_edgelist, delete_vertices=True)\n",
    "#     node_to_cluster = pd.Series(enx.get_undirected_igraph_connected_components(graph_clustered))\n",
    "#     del graph\n",
    "    \n",
    "#     # Calculate silhouette scores\n",
    "#     clustered_nodes = node_to_cluster.index\n",
    "#     index = clustered_nodes.map(lambda x: jaccard_distances.index.get_loc(x)).values\n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] Calculate silhouette scores using {len(index)} nodes\")\n",
    "#     dist = jaccard_distances.values[index,:][:,index]\n",
    "#     score = silhouette_score(dist, node_to_cluster.values, metric=\"precomputed\", sample_size=None, random_state=None) \n",
    "#     logger.info(f\"[n_neighbors = {n_neighbors}] [score = {score}] [n_nodes = {len(index)}] [n_edges = {graph_clustered.ecount()}]\")\n",
    "#     del dist\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "5acb36bc-e0f1-40de-8943-7edc21ac866e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pacmap\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from sklearn.datasets import make_blobs\n",
    "\n",
    "# Generate synthetic data with make_blobs\n",
    "n_samples = 100\n",
    "n_features = 10\n",
    "n_clusters = 3\n",
    "X, y = make_blobs(n_samples=n_samples, n_features=n_features, centers=n_clusters, random_state=42)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4f4b590e-a66d-430e-9b77-4c12fb39f045",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7fcdaf035460>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAgsAAAH5CAYAAAAP7X1BAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8ekN5oAAAACXBIWXMAAA9hAAAPYQGoP6dpAAAplklEQVR4nO3dfXjT9b3/8VeS0pSbNrXQWyilBbkpyK2jFnXqZBSGzLt51KkHmWOOuXkEnROngE4Gwvmhc4ejZ79Ncb/fJrodxc05jspQ5igguCpyo4BAi9BSCjRtgZQm3/NHIFhoP4Q26Telz8d15YLkm5u3uUrz9HsXh2VZlgAAAJrhtHsAAAAQ24gFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwirN7gNYKBALau3evEhMT5XA47B4HAIB2w7Is1dTUKCsrS05n8+sP2n0s7N27V9nZ2XaPAQBAu1VWVqZevXo1u7zdx0JiYqKk4H9oUlKSzdMAANB+eL1eZWdnhz5Lm9PuY+HkpoekpCRiAQCAFjjbZnx2cAQAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCEFIl6e+SPrd7EACIKcQCEHJIUoOkA3YPAgAxJc7uAYDYkSOpk6Tudg8CADGFWABCOikYDK1xWNJRSRmSHK0dCABiArEARIxfUokkS5JLUpqt0wBApLDPAhAxTklJCjZ4V5tnAYDIYc0COoBjkjZLSpaUd5b7HlVwU0KagmsHzoVD0shzfAwAxD5iAR1AlaRqSV5JuTLvS7BRUp2CgZEb/dEAoB0gFtABpCkYAEk6+06HSZKOSOoW7aEAoN0gFtABdJLUP8z7Djxx33PdnccnqULBtRddZF6DUX7i0k9ECYD2gFhAB7VbUo2CYRB/2rKW7Pe7TVLll673auJ5T9ql4L4RJ4MBAGIbR0OgAwooeErnSgX3Z4iEZAXXJCTq1MmdmtNXwU0jWRF6bQCILtYs4DxkSdqr4Id3ps7cHOBU8AO7VlKPZp6jVsE1AJlqfEbHgE6dR0EK7gjpUnBNQq8w50s9cQGA9oFYwHmoVtJnJ/7u0ZnnPLAU/GA3rVjbo+Cah2M6FQsBSetO3HbxidvWKxgLheKfE4DzFb/dcB7qrFNHPiSctiwg6QOd+sBv7uRJPRXcafHLmwqsE4+zJNXr1KYGKyJTA0CsIhZwHoqTNKqZZZaCOxee/LOzTq1h2K3gN08OVHDfg2GnPdalYGDUS7pAwRj5yonX458SgPMXv+HQwbgUDAmfpDJJmySNUHBNxC4F1zxUSspu5vGnH+oY7mmdLUmfKni+hyGS3OcyNADYiqMh0AElKrgfQo2CcXDkxO39FdyhMT0KrxmQtE/B8zBE6ggMAGgbrFlAB+VQcI1CnU7FQeaJi0nliUuugpswwuVS8GRNRxQMBg6bBNB+sGYBHcxxnTohU6KkDJ39FNBftl3BMzXubcFrp554LU8LHgsA9mHNAjqYMgVjYZ+kS1rw+D4KxkJGCx6bp7N/kRUAxB5iAR1MsoJrBVp6UqRwNlWYEAoA2h82Q+A85Fdwc8GeJpalSBqj4NEI57qjYUBSqaSDrZru1HNVKnhUBgDENmIB56FDCm5u2KbgORF04vqaE8sqTyz7WMGwMLEU3L+hQcHNDztOPK61J2LaI+kTSR+18nkAIPrYDIHzkEfBkyYl6NRZFvcpeBKmAwqenTFBwXMknK2XyyVtPXHfwQqukfCo9ZsTEk77EwBiF7GA81AnScNPu22ATp1sya3gdzmEw/mlP7squAkjEtIUDBr+CQKIffymQgfhUcsOWUxX8KyN0VgDYPoaawCIHcQCcFbhntIZAM5P7OCIDiIgvh0SAFqGWEAHcEjSKkkb7R4EANolYgEdwBGdOgQSAHCu2GcBHUCmgj/qiXYPAgDtErGADsCp6HztdHNqFTwPAztGAjg/sBkCMLJ0bjtGHpG0XtIHko5FZSIAaGusWQAa2aXgaZ3zJbkU/ODvLGmUwmtrp4JrFRxh3h8AYh+xADTyhYLfJ3FAUpKC3x1Rq+Chl+F8+Cfo1FkeOekSgPMDsQA0MkjBb5XsqeCH/cnvgziXfypEAoDzC7EANJJy4nJSWgufp0rSFgWjI7e1QwGAraK6UXXVqlWaNGmSsrKy5HA4tGzZskbLLcvSrFmzlJmZqc6dO2vs2LHatm1bNEcC2sghSccV/PIqAGjfohoLdXV1GjZsmBYvXtzk8gULFuiZZ57Rc889p7Vr16pr164qKirSsWPsRY72rreCaxTy7R4EAFotqpshJkyYoAkTJjS5zLIsPf3003rkkUd07bXXSpJ++9vfKj09XcuWLdMtt9wSzdGACNsrab+kHAW/ejpeUpakzxT81so+tk0GAK1l27FdO3fuVHl5ucaOHRu6zePxqKCgQMXFxc0+zufzyev1NroA9qqR9KmCmx62f+n2KgU3Q+xU8GgKAGifbIuF8vJySVJ6euMz66Wnp4eWNWXevHnyeDyhS3Z2dlTnBM4uXsFzMjgkffnnsYekDEl9xTkXALRn7e432MyZM1VdXR26lJWV2T0SOjy3pMskfVXBODipk4KHYva2YygAiBjbDp3MyAj+Uq2oqFBmZmbo9oqKCg0fPrzZx7ndbrnd7miPB5yjdtfdABA2237D5ebmKiMjQytWrAjd5vV6tXbtWhUWFto1FgAAOE1U1yzU1tZq+/ZTO3zt3LlTJSUlSklJUe/evXXffffpiSee0IUXXqjc3Fw9+uijysrK0nXXXRfNsQAAwDmIaiysX79eV111Vej6jBkzJEmTJ0/WkiVL9OCDD6qurk7f+973dPjwYV122WVavny5EhISojkWAAA4Bw7Lss7l+3djjtfrlcfjUXV1tZKSkuweBwCAdiPcz1D2ygIAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADCyPRbmzJkjh8PR6DJw4EC7xwIAACfE2T2AJA0ePFjvvPNO6HpcXEyMBQAAFCOxEBcXp4yMDLvHAAAATbB9M4Qkbdu2TVlZWcrLy9Ntt92m0tLSZu/r8/nk9XobXQAAQPTYHgsFBQVasmSJli9frmeffVY7d+7U5ZdfrpqamibvP2/ePHk8ntAlOzu7jScGAKBjcViWZdk9xJcdPnxYOTk5WrRoke66664zlvt8Pvl8vtB1r9er7OxsVVdXKykpqS1HBQCgXfN6vfJ4PGf9DI2JfRa+LDk5Wf3799f27dubXO52u+V2u9t4KgAAOi7bN0Ocrra2Vjt27FBmZqbdowAAAMVALDzwwAN67733tGvXLq1evVrXX3+9XC6Xbr31VrtHAwAAioHNEHv27NGtt96qqqoqpaam6rLLLtOaNWuUmppq92gAAEAxEAtLly61ewQAAGBg+2YIAAAQ24gFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYBQTsbB48WL16dNHCQkJKigo0Lp16+weCQAAnGB7LLz88suaMWOGZs+erQ8//FDDhg1TUVGR9u/fb/doAABAMRALixYt0tSpUzVlyhTl5+frueeeU5cuXfT888/bPRoAAJDNsVBfX68NGzZo7NixoducTqfGjh2r4uLiJh/j8/nk9XobXQAAQPTYGgsHDhyQ3+9Xenp6o9vT09NVXl7e5GPmzZsnj8cTumRnZ7fFqAAAdFi2b4Y4VzNnzlR1dXXoUlZWZvdIAACc1+LsfPEePXrI5XKpoqKi0e0VFRXKyMho8jFut1tut7stxgMAALJ5zUJ8fLxGjRqlFStWhG4LBAJasWKFCgsLbZwMAACcZOuaBUmaMWOGJk+erIsvvlijR4/W008/rbq6Ok2ZMsXu0QAAgGIgFm6++WZVVlZq1qxZKi8v1/Dhw7V8+fIzdnoEAAD2cFiWZdk9RGt4vV55PB5VV1crKSnJ7nEAAGg3wv0MbXdHQwAAgLZFLAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIhRdbU+1XiP2T0GsQAAQCyq9fo0Y+prmn7Xq6qsqLV1ljhbXx0AADSpocGv48f9kmXpeL3f1lmIBQAAYlByShf9/BeT1NDgV1a2x9ZZiAUAAGJURs8ku0eQxD4LAADEhIaGgGprfHaP0SRiAQCAGPD4g3/VPf/6irZsLLd7lDMQCwAAtIE9pYe1cM47en/lDknS+uJSvbb0I9X7GiRJB/bXSpb0j5Wf6x/vfq6jR+rtHLcRYgEAgDbwj7/t0Ccl+7Rs6cdqOO7X4oWr9PrLH2vd6t2SpNGX5UiSiv++U7/6xT/00gsb7By3EXZwBACgDVxZdKGqDtRp9KV9FNfJpUuvytOu7VUakJ8uSZpwbb5qqn0KBAL6cO0e9cpJtnfgL3FYlmXZPURreL1eeTweVVdXKykpNvYaBQCgJSzL0tt/2SrvoWO6/tvD5HI59cHq3dq7p1rfuH6wOnVyRfT1wv0MZc0CAAAxoqqyTr/79Xo5HFL+sAz1G5CqxQtXSZLS0hNVeEWuLXMRCwAAxIiU7l1UcGmOamp86tO3uzrFu1T41VyV7jqkCwel2jYXmyEAAOigwv0M5WgIAABgRCwAAAAjYgEAABgRCwAAtJK3+pgaGgJ2jxE1xAIAAK2w8Z979aPJf9ATDy23e5SoIRYAAGgF7+FjcjgcOnTwSIufw+8PxPSaCc6zAABAK4y5MlfJKZ2V2bNlh+/X1dbr4Xv/JH9DQHOfmSRPcucIT9h6rFkAAKCF/vyHjXpw2jLFu11K6dG1Rc9xpM6n6kNHVVvjU031sQhPGBmsWQAAoAWWLtmgt/68RX6/pY/Wf6ELB6a16HlS0xP10BPjFPAH1CvngghPGRnEAgAALfDe29vl91sadFGGxk0a1KrnGjg4PUJTRQexAABAC9zz48v16ab9mnBdvrp0jT9jeVVlnXy+BmX18jS6PeAP6OUXP5Qk/cvkkXK5Yn+PAGIBAIAWGDI8S0OGZzW5rK62Xg/98HX5GwKavfAbyslLCS0r231Yy/+0RZJUeEWu+vTt3ibztgaxAABAhDmdksvllL8hoLhOjdcc9OydrCu+3k+SzmkfhbJdh5ToSVDyBW1/tATfOgkAQBTUen2qP+5XSvcurX6uzR/v04LZ76hrN7d+8cK3FBcXmU0X4X6GsmYBAIAo6JbkjthzxcW5Tvxpz/4NxAIAAG2gfK9XlmUps6fn7HeW9O5b21S686C+dfsI9c9P08LnrlfXbvG2BAOxAABAlFVV1unhe/8sSXpy8TeVmp5ovH9DQ0BLnl0jS1Kfvt311bH9lJrerQ0mbVrsH68BAICNtn9aqdXvfa5AoOW7+LlcDrlcDjmdDrlObFIwiYtz6pobh2jw0AwNHdn0ERdtiTULAAA0o77er3k/fUt+f0CdOrn0lTE5YT/W7w+EzqGQnNJF/+dXN0iSkjwJYT3+W7ePOPeBo4Q1CwAANCMuzqk+/brL5XLqT69s1IH9tZKkZS9/rJ89tFwV+2qafNx7b2/TXTf9Xn/83T9DtyV5EsIOhVhDLAAA0Ayn06GH546TPxBQ6a5D+qRknyTpzVc3afvWSpV8sKfJx+3+/KCsgKXPPzvQluNGDZshAAAwcLmcuvu+y7Rre5UKLgtuhrjrR4X6dPN+XXplXpOP+dbtI5STl6Kho3oan/vwwSMq31ejAflpcjgcEZ89UjgpEwAANpn+3f/WoYNH9b1/G6MxVzQdHtEU7mcomyEAALBJYlKCrIClrt0idwKnaGAzBAAANnn0yfGqq/EpOaX1p4SOJtYsAABgk06dXMZQ2PpJhX79y9Xa90V1G051JmIBAIAI2/jPvfrlk+9p9+cHz3rfqso6fbZlf5PLXnphvf6+Yof+umxzpEc8J8QCAAARtmzpR1pfXKq3/rzFeD+/P6BHp7+hn//0LZWsP/MwzPHfzFffAT102df6RmvUsLDPAgAAEXbNjUP0zpuf6msT+jd7n2NHj+upuSt19OhxWQFL9b6GM+5TeEWuCq/IjeaoYSEWAACIsBGjszVidLbxPnv3VGvrJxWh62vf360uXeM1ZLj93wVxOmIBAAAb5Pbrrlu/M0pluw7pUNURbVhTqg/XlelXS29Vp05n/7KptkQsAABgA4fDofHfzJckle/16slZb6t3bori4mJvd8LYmwgAgHZm00f79B8LVjV79EMgYOnV35fo1d+XNPlV1+mZiRp0UYaOHqlXbY0v2uOeM2IBAIBWevX3Jfpg9W79TzNHP5TuPKjXX9mo11/ZqNKdZwbF7s8P6h8rP9dnmyu1bWtltMc9Z2yGAACglSbeOETv/GWrvja+6aMfuqd21YD8NCV5EtSzd/IZyxc9sVKSNPwrPTV0BDs4AgBw3hk5OlsjDUc/vLB4jT7bsl9FkwY1ufNij9Suqj58VF+fOFBxMbZzo0QsAAAQdUmeBFmWlOhJaHL5T+cV6djRBnXtFt/Gk4WHr6gGACDKAgFLh6qOKKVHFzkcDrvHCQn3M5Q1CwAARJnT6VD31K52j9FiHA0BAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAICRrbHQp08fORyORpf58+fbORIAADhNnN0DPP7445o6dWroemJioo3TAACA09keC4mJicrIyAj7/j6fTz6fL3Td6/VGYywAAHCC7fsszJ8/X927d9eIESO0cOFCNTQ0GO8/b948eTye0CU7O7uNJgUAoGNyWJZl2fXiixYt0siRI5WSkqLVq1dr5syZmjJlihYtWtTsY5pas5Cdna3q6molJSW1xdgAAJwXvF6vPB7PWT9DIx4LDz30kJ588knjfbZs2aKBAweecfvzzz+vu+++W7W1tXK73WG9Xrj/oQAAoDHbYqGyslJVVVXG++Tl5Sk+Pv6M2zdt2qQhQ4Zo69atGjBgQFivRywAANAy4X6GRnwHx9TUVKWmprbosSUlJXI6nUpLS4vwVAAAoKVsOxqiuLhYa9eu1VVXXaXExEQVFxdr+vTpuv3223XBBRfYNRYAADiNbbHgdru1dOlSzZkzRz6fT7m5uZo+fbpmzJhh10gAAKAJtsXCyJEjtWbNGrteHgAAhMn28ywAAIDYRiwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxEKEWZalw5t3KXC8we5RAACICGIhwj6a+zu9NuQu/X3KArtHAQAgIoiFCAv4jksOh/zH6iVJx2uO6M0r79Pb1zysgN9v83QAAJy7OLsHOB80HKvX5qf+qJThfTV8zr+q18QCpQzrK0n64q0PVLFqoySp/N0SZV09ys5RAQA4Z8RCBOx86W/a8Mjzcsa5dMeRN5V2SX5oWVxil9Dfj9f57BgPAIBWIRYiIO2yIUrsm6m0S/LldLkaLWuoPRr6e0L3pLYeDQCAVnNYlmXZPURreL1eeTweVVdXKykp9j6MLcvSZ8//VW5PN/X51lftHgcAgJBwP0NZsxBlDodDA+76ht1jAADQYhwNAQAAjIiFNuA76FXdF5V2jwEAQIsQC1F2vOaI/tjvDv0h7zYd+mRn6PaGoz599ps35d3+hY3TAQBwduyzEGVWICB//XHJkvz1x1VVsl2f/tcbcric2vrs6/IM7K0bNr1g95gAADSLWIiyeE833bBliRpqjyp5UI7evuZh7XlzrVIvyZfD5VJS/152jwgAgBGbIdpAt+w0JQ/KkSRd+J0JumBongb98Do541za85e1OrD+U5snBACgeaxZaGN9brhcfW64XMdrj2pt1wTVV9fJ+/k+dR/VXw6Hw+7xAAA4A2sWbNKpW2eNWz5fCenJeu/bT+jzl/5m90gAADSJNQttyAoEtO7+Z9VQd0zDfnqb/nLpvQrUN0iS4jq7bZ4OAICmEQttyLvtC23+xauS06Hsa8fI1dkth9OpcX+dp4wrhts9HgAATSIW2lDShT01eMZNaqg7qp5jR+lfdv1egQa/Erp77B4NAIBmEQttyOF0avS/fz903eWOt3EaAADCww6OAADAiFhoQ5UfbNUHP/kV3xMBAGhX2AzRhlbfvUgHS3aooe6YCv/jXrvHAQAgLKxZaEMX3jleiX2zlHP9ZXaPAgBA2Fiz0Iby771B6V8dqoMf7VCgwS9nnMvukQAAOCvWLLSx5V+7X+9PWaDXhnxHR8oP2j0OAABnRSy0sR6jB0qSvJ/t0b4VH9o8DQAAZ8dmiDZWtPxJffp//6LqLbvV+7pL7R4HAICzYs1CGD548L/0+9Trtfdv/4zI8w2YOlGjF/1Anbp2lu9QjTYuWKqDGz+PyHMDABBpxEIYdi97X76DNSp/tyTiz73xyaVaP/PXeu/bT0T8uQEAiAQ2Q5zGsiy9/50Fqt29X1f9YZYSunt01Suztfet9Rpw9zURe42SOS/Kd7hWWV8fJfdv3lTva9kkAQCITcTCaXwHqrX9xbfkcDpVWbxZ2dcUqvvwfuo+vF/EXqN2V7lKfvb/JIdDva+9VN+ufC1izw0AQKQRC6dJSE1W4X/ep7qy/coad3FUXqNbTroG/uCbqj9Uq7RLBqnui0ptWvRH5dz4VaWPGRyV1wQAoKUclmVZdg/RGl6vVx6PR9XV1UpKSrJ7nBZZe99ibX7mVSUP7qPrN/7G7nEAAB1EuJ+h7OAYA3JuuFyegdkaMHWi3aMAAHAG1iycxeEtu7XqjnnKnjRGI2b/a8SfHwAAu7BmIULK/lysqn9u0+Zf/LfdowAAYIuoxcLcuXM1ZswYdenSRcnJyU3ep7S0VBMnTlSXLl2UlpamH//4x2poaIjWSC1y4V0TlH/vjbp8yU8i9pyHPtkp3+HaiD0fAADRFLVYqK+v10033aRp06Y1udzv92vixImqr6/X6tWr9eKLL2rJkiWaNWtWtEZqkYTuHhU89QP1/uaYs97XsiwV//AZ/fXK6U1+SZTvUI02//JVLRv2Xb1xyT3RGBcAgIiL2qGTjz32mCRpyZIlTS5/6623tHnzZr3zzjtKT0/X8OHD9bOf/Uw/+clPNGfOHMXHx0drtKhpOHJMW//zdcnh0M6XV2rg9yfJ5Q7+d1iBgJYN/a6OfHFAkuRwOuwcFQCAsNm2z0JxcbEuuugipaenh24rKiqS1+vVpk2bmn2cz+eT1+ttdLHDjt+9o3UPPKfjNUdCt3Xq2lmX/voB9Rr/Fa2b/p/6n6JTmy4sy1KgwS9J6nvHWF2zZnGbzwwAQEvYFgvl5eWNQkFS6Hp5eXmzj5s3b548Hk/okp2dHdU5mxJo8Ovvdz6pTU/9QTv/8F6jZf2/M0HZ1xRKDoca6o6Gbne6XBr28Lclh7TjdysU19nd1mMDANAi5xQLDz30kBwOh/GydevWaM0qSZo5c6aqq6tDl7Kysqi+niQdrzmiP138fb120Xd0eGupnHEu5f/oeqVekq+sr4864/4Dvj9JE9//hUY/dY/+duNslb25VpLU51tXKOvqkRr6k1vl7MTJMwEA7cM5fWLdf//9uvPOO433ycvLC+u5MjIytG7duka3VVRUhJY1x+12y+1u2/8rr9lVrqoPt0mSXh/xPd1c9rJGL/pBs/d3OBxKKxys97/779r92vuqLduv7G8UqEtmdxW9tbDRfas/LdOxKi+neQYAxKxzioXU1FSlpqZG5IULCws1d+5c7d+/X2lpaZKkt99+W0lJScrPz4/Ia0TKBUNyNWr+VH3409/IGecKe+fEgdO+qaPlB9W/mTMzHq85omXDp8ryBzR+xb8r4/KhkoL7N9SV7lfX3mlyONgREgBgr6itCy8tLdXBgwdVWloqv9+vkpISSVK/fv3UrVs3jRs3Tvn5+brjjju0YMEClZeX65FHHtE999zT5msOzsbhcGjog7co79avyRUfJ3dKeGeK7DGqv77+xs+bf95OcUro4dGx/YcaPec/Zy/RR0/8f+X/2w0qeIpDLAEA9opaLMyaNUsvvvhi6PqIESMkSStXrtSVV14pl8ulN954Q9OmTVNhYaG6du2qyZMn6/HHH4/WSK3WLTstos8XlxCvGz/7rfy+49rw099o9x9X6eplj8t3wCs5HTpWWR3R1wMAoCX4bogYsTTrJh2tOKSL531Xg+69QRV//1jplw5RXJcEu0cDAJynwv0MZZf8GHH1sp+pfNXHGjhtkuIS4tXz6xfbPRIAAJKIhZiROnqgUkcPtHsMAADOwLdOAgAAI2IBAAAYEQsAAMCIWAAAAEbEAgAAMCIWAACAEbEAAACMiAUAAGBELAAAACNiAQAAGBELAADAiFgAAABGxAIAADAiFgAAgBGxAAAAjIgFAABgRCwAAACjOLsHaC3LsiRJXq/X5kkAAGhfTn52nvwsbU67j4WamhpJUnZ2ts2TAADQPtXU1Mjj8TS73GGdLSdiXCAQ0N69e5WYmCiHw2H3OFHn9XqVnZ2tsrIyJSUl2T1OzOJ9Cg/vU3h4n8LHexWeWHmfLMtSTU2NsrKy5HQ2v2dCu1+z4HQ61atXL7vHaHNJSUn8QwwD71N4eJ/Cw/sUPt6r8MTC+2Rao3ASOzgCAAAjYgEAABgRC+2M2+3W7Nmz5Xa77R4lpvE+hYf3KTy8T+HjvQpPe3uf2v0OjgAAILpYswAAAIyIBQAAYEQsAAAAI2IBAAAYEQsAAMCIWGhH5s6dqzFjxqhLly5KTk5u8j6lpaWaOHGiunTporS0NP34xz9WQ0ND2w4aY/r06SOHw9HoMn/+fLvHigmLFy9Wnz59lJCQoIKCAq1bt87ukWLKnDlzzvjZGThwoN1j2W7VqlWaNGmSsrKy5HA4tGzZskbLLcvSrFmzlJmZqc6dO2vs2LHatm2bPcPa7Gzv1Z133nnGz9j48ePtGdaAWGhH6uvrddNNN2natGlNLvf7/Zo4caLq6+u1evVqvfjii1qyZIlmzZrVxpPGnscff1z79u0LXX70ox/ZPZLtXn75Zc2YMUOzZ8/Whx9+qGHDhqmoqEj79++3e7SYMnjw4EY/O++//77dI9murq5Ow4YN0+LFi5tcvmDBAj3zzDN67rnntHbtWnXt2lVFRUU6duxYG09qv7O9V5I0fvz4Rj9jL730UhtOGCYL7c4LL7xgeTyeM25/8803LafTaZWXl4due/bZZ62kpCTL5/O14YSxJScnx3rqqafsHiPmjB492rrnnntC1/1+v5WVlWXNmzfPxqliy+zZs61hw4bZPUZMk2S99tproeuBQMDKyMiwFi5cGLrt8OHDltvttl566SUbJowdp79XlmVZkydPtq699lpb5jkXrFk4jxQXF+uiiy5Senp66LaioiJ5vV5t2rTJxsnsN3/+fHXv3l0jRozQwoULO/ymmfr6em3YsEFjx44N3eZ0OjV27FgVFxfbOFns2bZtm7KyspSXl6fbbrtNpaWldo8U03bu3Kny8vJGP1sej0cFBQX8bDXj3XffVVpamgYMGKBp06apqqrK7pHO0O6/dRKnlJeXNwoFSaHr5eXldowUE+69916NHDlSKSkpWr16tWbOnKl9+/Zp0aJFdo9mmwMHDsjv9zf587J161abpoo9BQUFWrJkiQYMGKB9+/bpscce0+WXX65PPvlEiYmJdo8Xk07+rmnqZ6sj/x5qzvjx43XDDTcoNzdXO3bs0MMPP6wJEyaouLhYLpfL7vFCiAWbPfTQQ3ryySeN99myZQs7VZ3mXN63GTNmhG4bOnSo4uPjdffdd2vevHnt5rzssMeECRNCfx86dKgKCgqUk5OjV155RXfddZeNk+F8ccstt4T+ftFFF2no0KHq27ev3n33XV199dU2TtYYsWCz+++/X3feeafxPnl5eWE9V0ZGxhl7s1dUVISWnU9a874VFBSooaFBu3bt0oABA6IwXezr0aOHXC5X6OfjpIqKivPuZyWSkpOT1b9/f23fvt3uUWLWyZ+fiooKZWZmhm6vqKjQ8OHDbZqq/cjLy1OPHj20fft2YgGnpKamKjU1NSLPVVhYqLlz52r//v1KS0uTJL399ttKSkpSfn5+RF4jVrTmfSspKZHT6Qy9Rx1RfHy8Ro0apRUrVui6666TJAUCAa1YsUI//OEP7R0uhtXW1mrHjh2644477B4lZuXm5iojI0MrVqwIxYHX69XatWubPZILp+zZs0dVVVWNQisWEAvtSGlpqQ4ePKjS0lL5/X6VlJRIkvr166du3bpp3Lhxys/P1x133KEFCxaovLxcjzzyiO65554Ou7q9uLhYa9eu1VVXXaXExEQVFxdr+vTpuv3223XBBRfYPZ6tZsyYocmTJ+viiy/W6NGj9fTTT6uurk5Tpkyxe7SY8cADD2jSpEnKycnR3r17NXv2bLlcLt166612j2ar2traRmtXdu7cqZKSEqWkpKh3796677779MQTT+jCCy9Ubm6uHn30UWVlZYXCtCMxvVcpKSl67LHHdOONNyojI0M7duzQgw8+qH79+qmoqMjGqZtg9+EYCN/kyZMtSWdcVq5cGbrPrl27rAkTJlidO3e2evToYd1///3W8ePH7RvaZhs2bLAKCgosj8djJSQkWIMGDbJ+/vOfW8eOHbN7tJjwy1/+0urdu7cVHx9vjR492lqzZo3dI8WUm2++2crMzLTi4+Otnj17WjfffLO1fft2u8ey3cqVK5v8XTR58mTLsoKHTz766KNWenq65Xa7rauvvtr69NNP7R3aJqb36siRI9a4ceOs1NRUq1OnTlZOTo41derURoe/xwqHZVmWDY0CAADaCc6zAAAAjIgFAABgRCwAAAAjYgEAABgRCwAAwIhYAAAARsQCAAAwIhYAAIARsQAAAIyIBQAAYEQsAAAAo/8FnBHYFP7RAT0AAAAASUVORK5CYII=",
      "text/plain": [
       "<Figure size 600x600 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# initializing the pacmap instance\n",
    "# Setting n_neighbors to \"None\" leads to an automatic choice shown below in \"parameter\" section\n",
    "embedding = pacmap.PaCMAP(n_components=2, MN_ratio=0.5, FP_ratio=2.0) \n",
    "\n",
    "# fit the data (The index of transformed data corresponds to the index of the original data)\n",
    "X_transformed = embedding.fit_transform(X, init=\"pca\")\n",
    "\n",
    "# visualize the embedding\n",
    "fig, ax = plt.subplots(1, 1, figsize=(6, 6))\n",
    "ax.scatter(X_transformed[:, 0], X_transformed[:, 1], cmap=\"Spectral\", c=y, s=0.6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "0f5f7a8d-6c5d-4574-8dc7-69933dea60b8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[ 14.522537 ,  -2.4231167],\n",
       "       [ 14.760947 ,  -2.5488222],\n",
       "       [ 14.983173 ,  -2.7196372],\n",
       "       [-11.052016 ,  -9.761924 ],\n",
       "       [ -3.9366643,  11.281354 ],\n",
       "       [ 14.898793 ,  -2.8463619],\n",
       "       [ -4.5757318,  11.343062 ],\n",
       "       [ -4.5757327,  11.343061 ],\n",
       "       [ 15.224875 ,  -2.3603597],\n",
       "       [ -4.5757318,  11.343062 ],\n",
       "       [-11.074902 ,  -9.718701 ],\n",
       "       [-11.242196 ,  -9.837603 ],\n",
       "       [-11.177533 ,  -9.8698225],\n",
       "       [ -4.5757327,  11.343061 ],\n",
       "       [ 15.004303 ,  -2.685898 ],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [-11.031492 ,  -9.603006 ],\n",
       "       [ 14.527227 ,  -2.5120075],\n",
       "       [-11.146766 ,  -9.532513 ],\n",
       "       [ 14.586111 ,  -2.4440212],\n",
       "       [ 14.664809 ,  -2.9630704],\n",
       "       [ 14.570836 ,  -2.8299773],\n",
       "       [-11.131973 ,  -9.709583 ],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [ 14.781629 ,  -2.668578 ],\n",
       "       [ 14.586111 ,  -2.4440212],\n",
       "       [ 14.682197 ,  -2.7698903],\n",
       "       [ 14.527228 ,  -2.512007 ],\n",
       "       [ -4.338963 ,  11.258381 ],\n",
       "       [ 14.382084 ,  -2.947289 ],\n",
       "       [-11.149052 ,  -9.849696 ],\n",
       "       [-11.019303 ,  -9.777588 ],\n",
       "       [ -4.498994 ,  11.256969 ],\n",
       "       [ -4.5757327,  11.343061 ],\n",
       "       [ 14.462413 ,  -2.365124 ],\n",
       "       [-11.070702 ,  -9.921214 ],\n",
       "       [ -4.4434457,  11.277985 ],\n",
       "       [ -4.5757327,  11.343061 ],\n",
       "       [-10.972218 ,  -9.215215 ],\n",
       "       [ 14.597886 ,  -2.5831792],\n",
       "       [-11.17068  ,  -9.688327 ],\n",
       "       [ -4.5757318,  11.343062 ],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [ -4.5757327,  11.343062 ],\n",
       "       [ -4.525725 ,  11.3359785],\n",
       "       [-11.0301075,  -9.407917 ],\n",
       "       [-11.08996  ,  -9.542981 ],\n",
       "       [ -4.3969707,  11.3839855],\n",
       "       [ 14.527228 ,  -2.5120068],\n",
       "       [-11.149052 ,  -9.849696 ],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [ 15.042215 ,  -2.1122515],\n",
       "       [-11.031492 ,  -9.603005 ],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [ -4.5757327,  11.343061 ],\n",
       "       [-10.99142  ,  -9.412035 ],\n",
       "       [ -4.575732 ,  11.343061 ],\n",
       "       [ 14.948415 ,  -2.2883887],\n",
       "       [-11.131973 ,  -9.709583 ],\n",
       "       [ 14.813449 ,  -2.2757967],\n",
       "       [ 14.896968 ,  -2.5365233],\n",
       "       [ -4.575732 ,  11.343061 ],\n",
       "       [ 14.688192 ,  -2.3988545],\n",
       "       [ -4.5757327,  11.343061 ],\n",
       "       [ -4.575732 ,  11.343062 ],\n",
       "       [-10.975015 ,  -9.6199045],\n",
       "       [ 14.600496 ,  -2.2612362],\n",
       "       [-10.994716 ,  -9.628643 ],\n",
       "       [ -4.575732 ,  11.343062 ],\n",
       "       [ -4.575732 ,  11.343061 ],\n",
       "       [ -4.5757327,  11.343062 ],\n",
       "       [ 14.497844 ,  -2.6919143],\n",
       "       [ -4.3973236,  11.379095 ],\n",
       "       [ 14.586111 ,  -2.4440215],\n",
       "       [ -4.572997 ,  11.173487 ],\n",
       "       [ 14.663312 ,  -2.2807524],\n",
       "       [-11.182696 ,  -9.860404 ],\n",
       "       [ 14.888292 ,  -2.0951564],\n",
       "       [-11.150184 ,  -9.751297 ],\n",
       "       [-11.182696 ,  -9.860404 ],\n",
       "       [ 14.527228 ,  -2.512007 ],\n",
       "       [ -4.5757327,  11.343061 ],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [ -4.575732 ,  11.343062 ],\n",
       "       [ 14.78163  ,  -2.6685777],\n",
       "       [-11.074902 ,  -9.718701 ],\n",
       "       [ 14.813903 ,  -2.141097 ],\n",
       "       [ -4.52095  ,  11.18811  ],\n",
       "       [ -4.338963 ,  11.258381 ],\n",
       "       [ -4.3973236,  11.379095 ],\n",
       "       [ -4.4965143,  11.136625 ],\n",
       "       [ 14.92692  ,  -2.2226894],\n",
       "       [-11.070702 ,  -9.921215 ],\n",
       "       [-11.019303 ,  -9.777587 ],\n",
       "       [ -4.3627353,  11.126607 ],\n",
       "       [ -4.575732 ,  11.343061 ],\n",
       "       [ 14.597885 ,  -2.5831792],\n",
       "       [ -4.5757318,  11.343062 ],\n",
       "       [-11.052016 ,  -9.761924 ]], dtype=float32)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "embedding.transform(X*2, basis=X)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "9e73dacd-9faa-4099-8577-c5197b07f10c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(700, 26) (300, 26)\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.datasets import make_classification\n",
    "from sklearn.model_selection import train_test_split\n",
    "from metabolic_niche_space.neighbors import KNeighborsKernel\n",
    "from metabolic_niche_space.manifold import DiffusionMapEmbedding # Shortcut: from datafold.dynfold import DiffusionMaps\n",
    "\n",
    "# Create dataset\n",
    "n_samples = 1000\n",
    "n_neighbors=int(np.sqrt(n_samples))\n",
    "X, y = make_classification(n_samples=n_samples, n_features=10, n_classes=2, n_clusters_per_class=1, random_state=0)\n",
    "X_training, X_testing, y_training, y_testing = train_test_split(X, y, test_size=0.3)\n",
    "\n",
    "# Build KNeighbors Kernel\n",
    "kernel = KNeighborsKernel(metric=\"euclidean\", n_neighbors=30)\n",
    "\n",
    "# Calculate Diffusion Maps using KNeighbors\n",
    "model = DiffusionMapEmbedding(\n",
    "    kernel=kernel, \n",
    "    n_eigenpairs=int(np.sqrt(X_training.shape[0])), # Upper bound\n",
    ") \n",
    "dmap_X = model.fit_transform(X_training)\n",
    "dmap_Y = model.transform(X_testing)\n",
    "\n",
    "# Shapes\n",
    "print(dmap_X.shape, dmap_Y.shape)\n",
    "# (700, 26) (300, 26)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "90391a09-35bb-4109-8696-d112695ce3bc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from sklearn.datasets import make_classification\n",
    "from sklearn.model_selection import train_test_split\n",
    "from metabolic_niche_space.neighbors import (\n",
    "    KNeighborsKernel,\n",
    "    pairwise_distances_kneighbors,\n",
    ")\n",
    "from metabolic_niche_space.manifold import DiffusionMapEmbedding # Shortcut: from datafold.dynfold import DiffusionMaps\n",
    "\n",
    "# Load real boolean data and create toy dataset\n",
    "n_observations = 500\n",
    "\n",
    "X_grouped = pd.read_csv(\"../test/X_grouped.tsv.gz\", sep=\"\\t\", index_col=0) > 0\n",
    "X_toy = X_grouped.iloc[:n_observations]\n",
    "X_training, X_testing = train_test_split(X_toy, test_size=0.3)\n",
    "\n",
    "# Precompute pairwise distances\n",
    "metric=\"jaccard\"\n",
    "\n",
    "distances = pairwise_distances_kneighbors(\n",
    "    X=X_training, \n",
    "    metric=metric, \n",
    "    n_jobs=-1, \n",
    "    redundant_form=True,\n",
    ")\n",
    "\n",
    "# Build KNeighbors Kernel with precomputed distances\n",
    "kernel = KNeighborsKernel(\n",
    "    metric=metric, \n",
    "    n_neighbors=50, \n",
    "    distance_matrix=distances.values,\n",
    ")\n",
    "\n",
    "# Calculate Diffusion Maps using KNeighbors\n",
    "model = DiffusionMapEmbedding(kernel=kernel, n_eigenpairs=int(np.log(X_training.shape[0]))) # Lower bound since it's sparse\n",
    "dmap_X = model.fit_transform(X_training)\n",
    "dmap_Y = model.transform(X_testing)\n",
    "dmap_X.shape, dmap_Y.shape\n",
    "# ((350, 5), (150, 5))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a247c26b-1acc-42d5-b825-f721af5a8f56",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'is_square_symmetric' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[18], line 38\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[38;5;66;03m# Bayesian-optimized KNN Leiden Clustering\u001b[39;00m\n\u001b[1;32m     29\u001b[0m clustering \u001b[38;5;241m=\u001b[39m KNeighborsLeidenClustering(\n\u001b[1;32m     30\u001b[0m     name\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mjaccard_similarity_clustering\u001b[39m\u001b[38;5;124m\"\u001b[39m, \n\u001b[1;32m     31\u001b[0m     feature_type\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mko\u001b[39m\u001b[38;5;124m\"\u001b[39m, \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     36\u001b[0m     n_jobs\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m,\n\u001b[1;32m     37\u001b[0m )\n\u001b[0;32m---> 38\u001b[0m \u001b[43mclustering\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfit\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdistances\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/SageMaker/environments/mns/lib/python3.9/site-packages/metabolic_niche_space/neighbors.py:719\u001b[0m, in \u001b[0;36mKNeighborsLeidenClustering.fit\u001b[0;34m(self, X, sampler, copy, **study_kws)\u001b[0m\n\u001b[1;32m    717\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mlen\u001b[39m(X\u001b[38;5;241m.\u001b[39mshape) \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m    718\u001b[0m     X \u001b[38;5;241m=\u001b[39m squareform(X)\n\u001b[0;32m--> 719\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[43mis_square_symmetric\u001b[49m(X):\n\u001b[1;32m    720\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIf X is precomputed, it must be square and symmetric\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    721\u001b[0m distance_matrix \u001b[38;5;241m=\u001b[39m X\n",
      "\u001b[0;31mNameError\u001b[0m: name 'is_square_symmetric' is not defined"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.model_selection import train_test_split\n",
    "from metabolic_niche_space.neighbors import (\n",
    "    KNeighborsLeidenClustering,\n",
    ")\n",
    "\n",
    "# Load real boolean data and create toy dataset\n",
    "n_observations = 500\n",
    "\n",
    "X_grouped = pd.read_csv(\"../test/X_grouped.tsv.gz\", sep=\"\\t\", index_col=0) > 0\n",
    "X_toy = X_grouped.iloc[:n_observations]\n",
    "X_training, X_testing = train_test_split(X_toy, test_size=0.3)\n",
    "\n",
    "# Precompute pairwise distances\n",
    "metric=\"jaccard\"\n",
    "\n",
    "distances = pairwise_distances_kneighbors(\n",
    "    X=X_training, \n",
    "    metric=metric, \n",
    "    n_jobs=-1, \n",
    "    redundant_form=True,\n",
    ")\n",
    "\n",
    "# Determine parameter range\n",
    "n = distances.shape[0]\n",
    "n_neighbors_params = [int, int(np.log(n)), int(np.sqrt(n)/2)]\n",
    "\n",
    "# Bayesian-optimized KNN Leiden Clustering\n",
    "clustering = KNeighborsLeidenClustering(\n",
    "    name=\"jaccard_similarity_clustering\", \n",
    "    feature_type=\"ko\", \n",
    "    observation_type=\"ani-cluster\", \n",
    "    class_type=\"LeidenCluster\", \n",
    "    n_neighbors=n_neighbors_params, \n",
    "    n_trials=5, \n",
    "    n_jobs=-1,\n",
    ")\n",
    "clustering.fit(distances)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e77cd4e3-ab2e-4dd2-acaf-c76c0aab5478",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "conda_mns",
   "language": "python",
   "name": "conda_mns"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.21"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
